### PR TITLE
analysis API implementations for kotlin source based symbols.

### PR DESCRIPTION
**Pull request by neetopia**
still quite WIP, but let's break the PR down into smaller ones when possible.

### CODE REVIEW

File: kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/KotlinSymbolProcessing.kt [hunk 1]
```diff
@@ -3,130 +3,72 @@ package com.google.devtools.ksp.impl
...
+    configureApplicationEnvironment(application)
 
-    val file = File("testData/api/hello.kt")
-    val text = FileUtil.loadFile(file, CharsetToolkit.UTF8, true).trim { it <= ' ' }
+    val files = kotlinSourceRoots
+            .sortedBy { Files.isSymbolicLink(it.toPath()) } // Get non-symbolic paths first
+            .flatMap { root -> root.walk().filter { it.isFile && it.extension == "kt" }.toList() }
+            .sortedBy { Files.isSymbolicLink(it.toPath()) } // This time is for .java files
+            .distinctBy { it.canonicalPath }
+    compilerConfiguration.addJavaSourceRoots(files)
```
Comment (ting-yuan) [inline kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/KotlinSymbolProcessing.kt:54]: Should it be `addKotlinSourceRoots`?

File: kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSPropertyDeclarationImpl.kt [hunk 2]
```diff
@@ -48,40 +44,59 @@ class KSPropertyDeclarationImpl(private val ktPropertySymbol: KtPropertySymbol)
...
+    }
+
+    override val location: Location by lazy {
+        ktPropertySymbol.psi?.toLocation() ?: NonExistLocation
+    }
+
     override val parent: KSNode? by lazy {
+        analyzeWithSymbolAsContext(ktPropertySymbol) {
+            ktPropertySymbol.getContainingSymbol()
+        }
```
Comment (ting-yuan) [inline kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSPropertyDeclarationImpl.kt:85]: Line 83 to 85 look redundant.

**Review by ting-yuan — COMMENTED** [review]
Two general comments:
1. Can some of the implementations, e.g., name, fqn, doc string, origin, location be shared?
2. For those implementations by psi, what would you do for libraries and java? We may need a list of them and try to see if they can be implemented in analysis api, along with other missing elements.

File: kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSClassDeclarationImpl.kt [hunk 3]
```diff
@@ -61,29 +51,47 @@ class KSClassDeclarationImpl(private val ktNamedClassOrObjectSymbol: KtNamedClas
     }
 
     override fun asStarProjectedType(): KSType {
-        TODO("Not yet implemented")
+        return analyzeWithSymbolAsContext(ktNamedClassOrObjectSymbol) {
+            ktNamedClassOrObjectSymbol.buildSelfClassType()
```
[the remark this answers is not in the document]
Comment (neetopia) [reply kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSClassDeclarationImpl.kt:55]: actually might need more attention on this, I will remove this implementation for now.

File: kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSAnnotationImpl.kt [hunk 4]
```diff
@@ -1,15 +1,8 @@
...
-import com.google.devtools.ksp.symbol.KSName
-import com.google.devtools.ksp.symbol.KSNode
-import com.google.devtools.ksp.symbol.KSTypeReference
-import com.google.devtools.ksp.symbol.KSValueArgument
-import com.google.devtools.ksp.symbol.KSVisitor
-import com.google.devtools.ksp.symbol.Location
-import com.google.devtools.ksp.symbol.Origin
+import com.google.devtools.ksp.symbol.*
 import org.jetbrains.kotlin.analysis.api.annotations.KtAnnotationApplication
+import org.jetbrains.kotlin.descriptors.annotations.AnnotationUseSiteTarget.*
```
[the remark this answers is not in the document]
Comment (neetopia) [reply kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSAnnotationImpl.kt:5]: there are several other places where descriptor classes are used, i will follow up with JB to see what's the plan.

**Review by ting-yuan — APPROVED** [review]

**Pull request merged**